lib/lz4opt.h

```diff
@@ -50,317 +44,289 @@ typedef struct {
-        if (!match_num) { ip++; continue; }
+        LZ4HC_match_t const firstMatch = LZ4HC_HashChain_GetAllMatches(ctx, ip, matchlimit, MINMATCH-1, nbSearches);
+        if (firstMatch.len==0) { ip++; continue; }
 
-        if ((size_t)matches[match_num-1].len > sufficient_len) {
+        if ((size_t)firstMatch.len > sufficient_len) {
             /* good enough solution : immediate encoding */
-            best_mlen = matches[match_num-1].len;
-            best_off = matches[match_num-1].off;
-            cur = 0;
-            last_pos = 1;
-            goto encode;
+            int const firstML = firstMatch.len;
+            const BYTE* const matchPos = ip - firstMatch.off;
+            if ( LZ4HC_encodeSequence(&ip, &op, &anchor, firstML, matchPos, limit, oend) )   /* updates ip, op and anchor */
+                return 0;  /* error */
+            continue;
         }
 
-        /* set prices using matches at position = 0 */
-        {   size_t matchNb;
-            for (matchNb = 0; matchNb < match_num; matchNb++) {
-                size_t mlen = (matchNb>0) ? (size_t)matches[matchNb-1].len+1 : MINMATCH;
-                best_mlen = matches[matchNb].len;   /* necessarily < sufficient_len < LZ4_OPT_NUM */
-                for ( ; mlen <= best_mlen ; mlen++) {
-                    size_t const cost = LZ4HC_sequencePrice(llen, mlen) - LZ4HC_literalsPrice(llen);
-                    SET_PRICE(mlen, mlen, matches[matchNb].off, 0, cost);   /* updates last_pos and opt[pos] */
-        }   }   }
-
-        if (last_pos < MINMATCH) { ip++; continue; }  /* note : on clang at least, this test improves performance */
+        /* set prices for first positions (literals) */
+        {   int rPos;
+            for (rPos = 0 ; rPos < MINMATCH ; rPos++) {
+                int const cost = LZ4HC_literalsPrice(llen + rPos);
+                opt[rPos].mlen = 1;
+                opt[rPos].off = 0;
+                opt[rPos].litlen = llen + rPos;
+                opt[rPos].price = cost;
+                DEBUGLOG(7, "rPos:%3i => price:%3i (litlen=%i) -- initial setup",
+                            rPos, cost, opt[rPos].litlen);
+        }   }
+        /* set prices using initial match */
+        {   int mlen = MINMATCH;
+            int const matchML = firstMatch.len;   /* necessarily < sufficient_len < LZ4_OPT_NUM */
+            int const offset = firstMatch.off;
+            assert(matchML < LZ4_OPT_NUM);
+            for ( ; mlen <= matchML ; mlen++) {
+                int const cost = LZ4HC_sequencePrice(llen, mlen);
+                opt[mlen].mlen = mlen;
+                opt[mlen].off = offset;
+                opt[mlen].litlen = llen;
+                opt[mlen].price = cost;
+                DEBUGLOG(7, "rPos:%3i => price:%3i (matchlen=%i) -- initial setup",
+                            mlen, cost, mlen);
+        }   }
+        last_match_pos = firstMatch.len;
+        {   int addLit;
+            for (addLit = 1; addLit <= 3; addLit ++) {
```

Comment:
Should this be `< MINMATCH`, or is it for a different reason?

Reply:
It's for a different reason.

The purpose of this code is to fill the cost for the next 3 positions after current best match.
Since we have no more match that far, these positions are filled with literals.

So why 3 ?
Well, this part is a bit more complex to explain.

Let's assume there is an initial match of length 17 (cost 4).
Next 3 positions 18, 19, 20 are filled with literals (respective cost 5, 6 and 7).
Let's now assume another match is found, at position 14 (cost 3).
At position 18, this match is not interesting, because total cost is 3+3=6, which is > cost(18)==5.
It starts to be interesting at position 19, as here cost(19)==6, so it's equal, and in this case, it's always beneficial to use the match (from a compression ratio perspective, decoding speed would prefer the literals).  That also means this second match must have a length >= 5 to be interesting. If its length is only 4, it's not an interesting match.

If the next 3 positions were not pre-filled with literal cost, all I would know is that the new match can reach position 18 > 17. In absence of competition, should it be selected blindly for position 18, it would be a bad decision, leading to worse cost.
Now, it's also possible to "calculate the cost of previous match + literal" on the fly, but for simplicity reason, I found it easier to simply pre-fill these 3 slots (and keep the `last_match_pos` to the real end position of the last useful match).

Pre-filling beyond 3 slots is not useful, because at such distance, a new match is necessarily better than 4+ literals.

So that's the reason for number `3`, and it is present in several places.
Maybe I could use a constant instead, in order to name it properly.

Reply:
Added constant `TRAILING_LITERALS`